### DISCUSSION TITLE
Profiling tests in Vitest

### DISCUSSION THREAD

**Discussion by timhn-bm** [vitest-dev/vitest#5800] [General]
Hi :wave: 

I am trying to do some **performance profiling on our tests** to identify bottlenecks and overall improve our testing strategy.

I am finding it quite difficult and cumbersome to do this profiling. I am wondering if I'm doing things wrong and if there is a better approach than mine with Vitest.

### Below is the ideal approach I would want to follow, but does not work with Vitest
This is what I did with Jest for instance, which worked. But not with Vitest

- Run vitest with` node --inspect`, with tests not running in parallel. For instance
`node --inspect ./node_modules/vitest/vitest.mjs <test-name> --no-file-parallelism --pool=threads --poolOptions.threads.singleThread `

- Use [ndb](https://github.com/GoogleChromeLabs/ndb) in the Dev Tools, run a performance profile and analyse the flamecharts 

Unfortunately, as we you can see in the screenshot, we mostly see **_idle_** **blocks, and nothing related to the tests execution.**
##### Test Profiling, from `node --inspect`
![image](https://github.com/vitest-dev/vitest/assets/142981345/6e605609-367b-4ca7-96aa-cf3d2d6b0284)

### Why is that happening ?
I don't fully understand why this is the case, but from what I read [here](https://github.com/vitest-dev/vitest/discussions/2916) and [here](https://vitest.dev/guide/features#threads), **Vitest run tests with worker threads or child processes.** What we see in the flamecharts is only the **execution of the main process** (and its main thread ?). This one is responsible for spawning the worker threads / child processes (depending on the pool option). Hence we don't see the actual test execution.

**NB:** please correct if anything is inexact, or simply wrong. This is what I've understood from Vitest execution so far.

### A solution ? 
The only solution that worked from me - described [here](https://github.com/vitest-dev/vitest/discussions/2916) - is to generate _cpuprofile_ files using `node --cpu-prof`. And then import these files in the Dev Tools. You can see the result in the screenshot below:

##### Test Profiling, from a cpuprofile file
![image](https://github.com/vitest-dev/vitest/assets/142981345/74e30371-6514-4583-8219-eaba12726fc8)

**NB:** 
- running `node --cpu-prof` generates multiple _cpuprofile_ files. You'd have to import the right one that contains the test execution
- I am not entirely sure why, but this technique only works with _threads_, and not _forks_

**Conclusion**: 
This solution works, but is very cumbersome. I'd love to be able to rerun my test in one save, and obtain the flamecharts immediately. Do you have recommendations to better do profiling ? Is there something that can be done at Vitest codebase level to be able to do this ?

Thanks a lot :heart:

**Comment by AriPerkkio** [comment]
Edit: There's now https://vitest.dev/guide/profiling-test-performance.html#test-runner

> running node --cpu-prof generates multiple cpuprofile files. You'd have to import the right one that contains the test execution

To generate only a single cpuprofile you could use following configuration:

```ts
import { defineConfig } from "vitest/config";

export default defineConfig({
  test: {
    pool: "threads",
    poolOptions: {
      threads: {
        singleThread: true,
        isolate: false,
        execArgv: ["--cpu-prof", "--cpu-prof-dir=./profiling"],
      },
    },
  },
});
```

> I am not entirely sure why, but this technique only works with threads, and not forks

I'm not sure why `--cpu-prof` doesn't work for `forks` (`node:child_process`) here. 🤔 

But it looks like `--prof` works:

```ts
import { defineConfig } from "vitest/config";

export default defineConfig({
  test: {
    pool: "forks",
    poolOptions: {
      forks: {
        singleFork: true,
        isolate: false,
        execArgv: ["--prof"],
      },
    },
  },
});
```

There's quite much on the post description but hopefully this helps.

**Reply by timhn-bm** [reply]
Thanks @AriPerkkio ! I hadn't been notified of your reply. I'll look into your suggestions

My main question still remains however. These solutions are not practical to profile tests (or anything else in fact) as they require to manually import a file in NDB each time. Ideally, we'd want to run some test and automatically generate flamecharts.

**Reply by AriPerkkio** [reply]
There is `vitest --inspect` and `vitest --inspect-brk` commands. Those connect the inspector session with test runners' sessions. Have you tried those commands? https://vitest.dev/guide/debugging.html#node-inspector-e-g-chrome-devtools

**Reply by OlliL** [reply]
Wonder how thats all supposed to work. With Google Dev Tools connecting to the inspector session, the execution gets halted. Starting performance recording won't let you resume the execution. Resuming the execution instead won't let you start the recording (in time). Rerunning the tests directly halts the execution on debugger attachment again.... So I wonder - how am I supposed to get performance statistics of some super slow running tests for further analysis?
Ndb is no option for me as it starts up, shows a window and directly closes without any further message.

**Reply by AriPerkkio** [reply]
@OlliL record the profile with `--cpu-prof` without opening any dev tools. Once the tests have completed, open the `*.profile` in dev tools and look at the results.

**Reply by OlliL** [reply]
That did the trick. Thanks a lot @AriPerkkio

**Reply by tankorsmash** [reply]
Could you share the specific command to include `--cpu-prof`? `NODE_OPTIONS` doesn't support it. Thank you!

edit: `node --cpu-prof .\node_modules\vitest\dist\cli.js`

**Reply by AriPerkkio** [reply]
@tankorsmash that won't record profile of test runners, only the main thread. See https://vitest.dev/guide/profiling-test-performance.html#test-runner

**Reply by tankorsmash** [reply]
@AriPerkkio Thank you. Part of my issue ended up being that the vitest docs defaulted to 'forks' for the pool config, but since I'm on vitest v1.3.0 `vitest run` uses the 'thread' pool by default, so my settings weren't getting used. Had I read more, I'd have noticed sooner. Thank you!

**Reply by AriPerkkio** [reply]
> but `vitest run` uses the 'thread' pool by default

Since v2 Vitest uses `pool: 'forks'` by default: 
- https://github.com/vitest-dev/vitest/pull/5047

**Reply by tankorsmash** [reply]
Ah, I'm on 1.3.0 still, perfect! Thanks
